[terra-show-hide] prop to focus on the element added

adoroshk opened the pull request:
### Summary
* A new `focusRef` prop was added to pass a ref to the element that should be focused after the full text is shown.
* A `<Focuser>` component has been added for cases when user needs to set focus to a text which is not normally focusable. 
* When read by screen readers, upon expanding the content the screen reader starts reading the content on the newly revealed content when the component is expanded.
* Documentation was updated with the examples and implementation guidance.
* A fall-back behavior for no `focusRef` provided by customers was preserved for passivity, however it doesn't provide full accessibility.
* `aria-controls` property has been added to the ShowHide button. It contains the id of the ShowHide content `<div>`

### Testing
This change was tested using:

- [x] WDIO
- [x] Jest
- [ ] Visual testing (please attach a screenshot or recording)
- [x] Other (please describe below)
- [ ] No tests are needed

The following specific cases has been tested with assistive technologies:
* Microsoft Edge + JAWS, Safari, Chrome, Firefox + VO: tested for being walk-able with TAB and VO + Right/Left Arrow.
* JAWS + Microsoft Edge: tested the show-hide content to be walk-able with Up/Down Arrow.
* Safari + VO: tested the content to be walk-able via Up/Down Arrow.

### Reviews

In addition to engineering reviews, this PR needs:
<!-- Please include the appropriate "Required" & "Ready" labels (ex. "UX Review Required" and "UX Review Ready").  -->
- [x] UX review
- [x] Accessibility review
- [ ] Functional review

**This PR resolves:**

UXPLATFORM-9213
UXPLATFORM-9055

sdadn commented on packages/terra-core-docs/CHANGELOG.md:
This looks out of date from main; can you rebase so that it's up to date?

adoroshk replied:
Added tests: https://github.com/cerner/terra-core/pull/3838/commits/b28173d16f2ede07d53f4c5bcfc6aa3b4221de16

sycombs commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -56,17 +62,33 @@ const ShowHide = (props) => {
+      }
     }
-  }, [isOpen]);
+  }, [isOpen, props.focusRef, contentRef]);
```
I think we can just use `focusRef` since we destructured the props above.

adoroshk replied:
Good catch, removed: https://github.com/cerner/terra-core/pull/3838/commits/ccf7a43bde31fc3eedf2b274cc4a2d5f9249f738

sycombs commented on packages/terra-show-hide/CHANGELOG.md:
```diff
@@ -3,7 +3,7 @@
 
 * Added
-  * Shift focus to newly revealed content for screen-reader users.
+  * New focusRef prop to shift focus to upon revealing the hidden content for screen-reader users.
```
Can we reword to:
```suggestion
  * Added `focusRef` prop to shift focus to the hidden content when it is revealed for screen-reader users.
```

adoroshk replied:
Good catch! Thank you for suggesting. Added here: https://github.com/cerner/terra-core/pull/3838/commits/848c29f87e0af5b43f6eaade4494fd8a82d0ba6e

sycombs commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -35,6 +35,12 @@ const propTypes = {
    * Allows parent to toggle the component. True for open and false for close.
    */
   isOpen: PropTypes.bool,
+  /**
+   * Ref to the first hidden child element. The element will get focus once the full content revealed. That allows the assistive technologies to start reading the hidden content from the place where it was cut off.
+   */
```
can we reword slightly:
```suggestion
  /**
   * Ref to the first hidden child element that will receive focus when the full content is revealed. This allows assistive technologies to start reading the hidden content from where it left off.
   */
```

adoroshk replied:
Good catch! Thank you for suggesting. Added here: https://github.com/cerner/terra-core/pull/3838/commits/848c29f87e0af5b43f6eaade4494fd8a82d0ba6e

sdadn commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -56,17 +62,33 @@ const ShowHide = (props) => {
+        element.setAttribute('tabIndex', '-1');
+        element.setAttribute('data-focus-styles-enabled', isOpen);
+        element.setAttribute('role', 'group');
+        element.focus();
```
Can this be unit tested to see if the correct element is getting focus?

adoroshk replied:
Added tests: Added tests: https://github.com/cerner/terra-core/pull/3838/commits/b28173d16f2ede07d53f4c5bcfc6aa3b4221de16

cm9361 commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -44,6 +51,42 @@ const defaultProps = {
+  element.addEventListener('blur', disableFocusStyles);
+  // try to focus the element.
+  element.focus();
+  // if element is not focusable, add atributes that make it focusable, then focus again.
+  if (document.activeElement !== element) {
+    element.setAttribute('tabIndex', '-1');
+    element.setAttribute('role', 'group');
+    element.addEventListener('blur', (event) => {
+      disableFocusStyles(event);
+      removeAttributes(event);
+    });
+    element.focus();
+  }
```
Is this the correct approach?  I would expect the consumer to do the work necessary to ensure that the element is focusable.  I don't believe the platform component should be making updates to the DOM of our consumers.

adoroshk replied:
This case is unique as we need to make a normally not intractable DOM element like <span> placed inside the <p> to get focus for a short period of time (once the full content was revealed) to be announced by an assistive technologies. Then, once user moves away from that DOM element, it should lose that focus, as it should not be able to get it back. The logic that has been implemented is following:
1. Try to set focus to the DOM element of interest.
2. Check if element got focus (meaning the DOM element is focus-able as is, like link). If yes, we do not touch the DOM element at all, the code block in question is omitted as `if (document.activeElement !== element)` statement returns `false`.
3. If no, we temporarily set the `tabindex` and `role` to the DOM element, but we also attach the event listener, which would remove them on blur event. After DOM element loses focus it gets back to the exact state it was.

Summarizing that, focus-able DOM elements stay intact in all cases, not focus-able DOM elements' attributes change their tab-index for a short period of time, then they got back to their initial state.

cm9361 replied:
If the element is not focusable, then you cannot set focus to it and would not get the style.  We should not have to add an event listener to the DOM directly, right?

adoroshk replied:
I might have misunderstood the question, but I got it as why we need the event listener here. Let me clarify "If the element is not focusable" part here. Let's assume the element is `<span>`. Initially it's not focusable. 
Once the content is revealed, it has to become focusable to get focus so that assistive technologies read it out, that is why `tabindex= -1` attribute is added. Now it's `<span tabindex="-1">`, and it is focusable. 
While `tabindex=-1` does not allow the element to get focus by tabbing, it still can get focus by clicking on the element with the mouse. It is a `<span>` element that does nothing on click, it should not get focus more than once upon opening the show-hide content. Hence, span should get back to its usual non-focusable self. 
Per that, at the moment when the 'tabindex=-1' is being added to the element, the event listener onBlur event is added as well. That even listener removes 'tabindex=-1' returning span to its initial state.

cm9361 commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -44,6 +51,42 @@ const defaultProps = {
+ * Focus element.
+ */
+const focusElement = (element) => {
+  element.setAttribute('data-focus-styles-enabled', 'true');
+  element.addEventListener('blur', disableFocusStyles);
```
Do we really need this logic to set the attribute?  We should be able to create CSS such that any descendant of the component that has focus gets the new style.

adoroshk replied:
This is a good question. I noticed, that some of terra components use styling on focus like this: `.button[data-focus-styles-enabled='true']:focus`, so I added that attribute for consistency, in order to not loose focus styling for these components. Let me do some more research as for how many components actually use that approach and get back with the update on focus styling approach.

adoroshk replied:
I double checked and I think it's safe to remove it as the usage of that attr was in terra-docs button, so I did: https://github.com/cerner/terra-core/pull/3838/commits/b87f7ad08f6658a55f44d86bab0143406f6ccae1

adoroshk commented on packages/terra-show-hide/src/ShowHide.module.scss:
```diff
@@ -14,4 +18,16 @@
+      outline: var(--terra-show-hide-focus-outline, 2px dashed #000);
+      outline-offset: var(--terra-show-hide-focus-outline-offset, 2px);
+    }
+  }
```
Good catch, at some point I was trying different styles and left it separated. Fixed here: https://github.com/cerner/terra-core/pull/3838/commits/1058492783f3c07cde274d84fd35bc6ab1ab161b

eawww commented:
Commenting for visibility. Discussions around the styling for this are ongoing. Hope to resolve soon.

mjpalazzo commented:
Great progress @adoroshk! Here are my observations:
Keyboard navigation with Chrome and Edge works well.
Screen reader testing with JAWS is generally good:
Used H key to get to header of first example. JAWS announced heading text and heading level.
Used Down Arrow to navigate text content. JAWS announced text as I pressed Down Arrow until the end of the visible text. 
Pressed Down Arrow and JAWS repeated announcement of unhidden text from " the urgency to grow..." This is a problem
Pressed Down Arrow until I got to the end of the unhidden text. JAWS announced the label, type and state, "Show More button collapsed." There was no visual focus indicator and JAWS did not announce how to interact with the button. It should announce "To activate press Enter."
Pressed Enter. Hidden text displayed with dashed border around it. JAWS announced the text as I pressed Down Arrow until end.
Pressed Down Arrow. JAWS announced "Hide button expanded." There was no visual focus indicator and JAWS did not announce how to interact with the button. It should announce "To activate press Enter."
Pressed Enter and text was hidden again. JAWS announced "Show More button collapsed." There was no visual focus indicator but this time JAWS announced how to interact with the button "To activate press Enter."
I tested the example with multiple children.
JAWS navigation and announcement behavior was same as previous example.
I pressed Show 2 More Sentences and the hidden text became unhidden. Dashed border displayed around first sentence.
Pressed Down Arrow. The dashed border did not move to second sentence. JAWS announced the text of the second sentence.
Pressed Down Arrow. JAWS announced Hide 2 more sentences button collapsed.
Pressed Enter. Unhidden text became hidden. JAWS immediately started reading next of next example. This was odd behavior that should be fixed.
Let me know if you want me to demonstrate these findings. I will do some testing with Safari and VoiceOver next.

adoroshk commented:
@mjpalazzo thank you for such a detailed description of testing, as always very appreciated.

**Used Down Arrow to navigate text content. JAWS announced text as I pressed Down Arrow until the end of the visible text.
Pressed Down Arrow and JAWS repeated announcement of unhidden text from " the urgency to grow..." This is a problem**
I have made some changes and don't observe this behavior anymore, unless it is presenting itself at some specific circumstances (I tried in edge). Hope this is fixed.

**Pressed Down Arrow until I got to the end of the unhidden text. JAWS announced the label, type and state, "Show More button collapsed." There was no visual focus indicator and JAWS did not announce how to interact with the button. It should announce "To activate press Enter."**
I tested with other buttons on the page, and buttons on Button component page. When navigated to the button with Down Arrow, JAWS didn't announce "To activate press Enter" for any of them and the visual focus indicator wasn't shown. It looks like it announces how to interact and focus indicator if I navigate to then by TAB, but not by Down Arrow. I assume it's an expected behavior. If not, please direct me to the example, so that I can check what is done differently.

**Pressed Enter. Unhidden text became hidden. JAWS immediately started reading next of next example. This was odd behavior that should be fixed.** - I did not observe that behavior, probably because of the same changes that I recently implemented, unless it is specific to the browser, I was testing in Edge. 

**VoiceOver was not able to navigate to the Show More button with the Down Arrow key.** I observe that behavior too. The problem is, that show-hide component uses toggle component inside, and when I tried toggle and toggle-button components, they've shown the same behavior. That behavior needs fixing within Toggle component.

cm9361 commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -94,25 +87,39 @@ const ShowHide = (props) => {
+        if (checkIfAddTabindex(focusRef?.current)) {
+          // not focusable element with no tabindex set, safe to temporary add tabindex that will later be removed on blur
+          focusRef?.current.setAttribute('tabIndex', '-1');
+          focusRef?.current.setAttribute('role', 'group');
```
Do we want to change the role of the focus ref component?  Couldn't that have negative accessibility impacts.  For example, what if the ref was a table or some other role?  We would now be blindly making it a group role.

adoroshk replied:
Makes sense, I'll check the role first and re-assign it only in case there was no role. Will add a commit per that and update you once it done.

adoroshk replied:
Added in this commit: https://github.com/cerner/terra-core/pull/3838/commits/170f7cbd0b5ff79272fb22d69c3b7d268398c352

cm9361 commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -55,29 +55,22 @@ const defaultProps = {
+  * element.tabIndex:  -1
+  * element.hasAttribute():  true
+  */
+const checkIfAddTabindex = (element) => element?.tabIndex === -1 && !element?.hasAttribute('tabindex');
```
We would only need to do this if the focusRef is passed in.  Can't we create a const value that lets us know whether to update the tabIndex.  The check should only be needed once?

adoroshk replied:
Correct, that is how it works now, as I call that method only one time, if the focusRef is passed in: https://github.com/cerner/terra-core/blob/6d1d09062f40aaba64069bfac4eba48e6d2d6dfe/packages/terra-show-hide/src/ShowHide.jsx#L99-L100

cm9361 commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -94,25 +87,39 @@ const ShowHide = (props) => {
 
   const theme = React.useContext(ThemeContext);
   const contentRef = React.useRef(null);
+  const [activeElement, setActiveElement] = React.useState(null);
```
This state variable is only for managing the state of the contentRef tab index.  It should be a Boolean value, right?

adoroshk replied:
My idea was to store an actual active element: contentRef element, focusRef element or null, if there is no active element at the moment. Boolean option has only 2 options and it is not enough for me to make a proper decision, as false would only indicate that contentRef is not an active element, saying nothing of if focusRef is currently active or there is no active element at all. I need to know an actual active element for blur, for setting tabindex, and it needs to have 3 options: contentRef element, focusRef element or no element at all, that is why boolean wouldn't work.

cm9361 replied:
Your state variable is only controlling whether we are setting the tabIndex on the content ref.  It is causing you to add unnecessary logic, as-is.

adoroshk replied:
It's not an unnecessary logic, it is the same logic that we need in case you suggested, just keeping track on active element for blur event, 3 states instead of 2, not much different. I changed it to look the way you want in this commit to resolve it, but it is pretty much the same: https://github.com/cerner/terra-core/pull/3838/commits/170f7cbd0b5ff79272fb22d69c3b7d268398c352

cm9361 commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -94,25 +87,39 @@ const ShowHide = (props) => {
+          // not focusable element with no tabindex set, safe to temporary add tabindex that will later be removed on blur
+          focusRef?.current.setAttribute('tabIndex', '-1');
+          focusRef?.current.setAttribute('role', 'group');
+          focusRef?.current.classList.add(focusRefClassName); // needed for onBlur cleanup
```
You want to avoid DOM manipulations, where possible, with React.  You should not have to add a class name.  Is this something you can manage with a reference variable?

adoroshk replied:
In this case we have no choice but manipulate DOM changing tabindex, and as we have to do so, I prefer the "marker" of it being manipulated being assigned the same way the tabindex was changed in case of asynchronicity between react element and DOM element. What I am trying to avoid is that I set a variable via React then change DOM attribute via setAttribute and at some point React state is not in sync with DOM, giving me a react variable indication it was manipulated when DOM element has already changed the attribute back (or the other way around). I believe that trying to sync a React reference variable with DOM element attribute state change is not 100% possible and the only way to be safe here is consistently attach the marker that tabIndex was changed the same way we are changing tabIndex.

cm9361 replied:
I am not understanding how they will get out of sync since React is controlling the rendering.  Maybe I am misunderstanding, but it seems like you just need to keep track of whether you need to remove the attributes.

adoroshk replied:
I'll try to rephrase what I meant. Technically, the React element and DOM element aren't the same, that is the exact reason why we don't want to manipulate the DOM elements by "getAttribute" from react - we can't be sure that at some point react element prop like tabIndex is not split second ahead or behind the element.getAttribute('tabindex'). As react is manipulating by virtual DOM rather than read DOM, if we resort to getAttribute tactics I am not sure we should use a react variable to keep track on that.

adoroshk replied:
Ok, as it's quite a grey area here, I'll look into adding a reference variable. Will update once have a commit.

adoroshk replied:
Variables to track if tabIndex and role were modified added: https://github.com/cerner/terra-core/pull/3838/commits/22a779db64ad198d7388aa83c1d11e1539b14033

cm9361 commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -94,25 +87,39 @@ const ShowHide = (props) => {
+          focusRef?.current.setAttribute('role', 'group');
+          focusRef?.current.classList.add(focusRefClassName); // needed for onBlur cleanup
+        }
+        newActiveElement = focusRef?.current;
```
You would only set the state for the contentRef.  There is no need to set if we are using the focusRef.

adoroshk replied:
No, I need to know the actual active element being contentRef, focusRef, or null. Here is the explanation: https://github.com/cerner/terra-core/pull/3838#discussion_r1263961154

adoroshk replied:
This also applies to that: https://github.com/cerner/terra-core/pull/3838#discussion_r1264094548

cm9361 commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -94,25 +87,39 @@ const ShowHide = (props) => {
+    setActiveElement(newActiveElement);
   }, [isOpen, focusRef, contentRef]);
 
+  /**
+  * Set focus to active element allows assistive technologies to read the newly revealed content.
+  */
+  React.useEffect(() => {
+    if (activeElement && isOpen) {
+      activeElement.focus();
+    }
+  }, [activeElement, isOpen]);
+
```
I don't believe you should need this additional useEffect hook.

adoroshk replied:
I need to set focus to the actual active element once it changes, I don't see how it can be done without the useEffect monitoring that change.

cm9361 replied:
Can't you set focus in the hook that is triggered by the isOpen prop change?

adoroshk replied:
I wouldn't work because that would trigger focus before tabIndex change to -1 in contentRef, meaning setting focus to not focusable element.

cm9361 commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -55,9 +81,45 @@ const ShowHide = (props) => {
   } = props;
 
+  const theme = React.useContext(ThemeContext);
+  const contentRef = React.useRef(null);
```
You can import these methods like this:
import React, { useContext, useRef, useCallback } from 'react';

Of course you would add all the ones that you require.

adoroshk replied:
Is there any benefit of importing them that way rather than use like **React.useContext**?

sdadn replied:
There is no programmatic advantage outside of not having to type "React" before every usage. As much as I prefer the old school way of typing the library name first, Charles' suggestion makes it consistent with the rest of Terra.

adoroshk replied:
I picked it up from theme context and kind of thought it is consistent with the rest of Terra the way it is: https://github.com/cerner/terra-core/blob/70ccf4fee8dafedd22ccdb4a224c2227bc7e764f/packages/terra-alert/src/Alert.jsx#L138, but I checked the rest of cases, and for useEffect the Charles's version is used everywhere, so changing for consistency: https://github.com/cerner/terra-core/pull/3838/commits/22a779db64ad198d7388aa83c1d11e1539b14033

sdadn replied:
Terra and inconsistency, name a better duo.

mjpalazzo commented:
@adoroshk - I retested with JAWS and some of the issues I reported previously are still present:
Used Down Arrow to navigate text content. At the end of visible text, I pressed Down Arrow again and focus went to the Show More button. JAWS announced button label, state. Press Enter, the hidden text became unhidden, and JAWS announced the unhidden text. When it got to the end I pressed Down Arrow again (screen reader user would not know when they get to the of the visible text, so it seems natural that they would continue to press Down Arrow. This caused JAWS to go back to the unhidden text, but not to the beginning of the unhidden text. It in the first example it went to the "the urgency to grow..." I recreated this behavior in other examples. This is an issue that needs to be corrected. The user is forced to hear partial repetition of the text previously announced. I will set up some time to demonstrate.

adoroshk commented:
@mjpalazzo Thank you for more testing!
**This caused JAWS to go back to the unhidden text, but not to the beginning of the unhidden text.**
I can observe it too. It looks like first JAWS announces the whole hidden text because of it getting focus. But when it stops, it basically stops on what JAWS considers a line, line that includes the beginning on the hidden text. On arrow Down it reads the next line (as it should). The problem is, I haven't found a way to influence JAWS reading the focused text in whole. As the component is quite unique, I can't found any examples similar to what we are truing to achieve. At this point it looks like that one can't be fixed. 

Do you by any chance know any examples with focusing text paragraphs that can be of help?

sdadn commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -44,8 +53,27 @@ const defaultProps = {
 };
 
+/**
+  * Returns true if element is NOT normally interactive element like button and has no negative tabindex added by user.
```
+1 on the well written documentation!

sdadn commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -55,9 +83,47 @@ const ShowHide = (props) => {
+  /**
+   * Upon showing hidden content, set activeElement to focusRef element if provided, othervise set it to content container.
+   */
+  useEffect(() => {
+    if (isOpen) {
+      if (focusRef?.current) {
+        if (checkIfAddTabindex(focusRef?.current)) {
+          // not focusable element with no tabindex set, safe to temporary add tabindex that will later be removed on blur
+          focusRef?.current.setAttribute('tabIndex', '-1');
+          isModified.current.tabIndex = true;
+          if (!focusRef?.current.hasAttribute('role')) {
+            focusRef?.current.setAttribute('role', 'group');
+            isModified.current.role = true;
+          }
+        }
+        focusRef?.current.focus();
+      } else if (contentRef?.current) {
+        setContainerIsActive(true);
+      }
+    }
+  }, [isOpen, focusRef, contentRef]);
```
Not a fan of having the whole function body be an `if` block. Can we change this to something like:

```js
if(!isOpen)
  return;

// rest of the function

```

This will reduce at least 1 nested `if`.

adoroshk replied:
Fixed in this commit: https://github.com/cerner/terra-core/pull/3838/commits/4d398feeb8d9d23ccb3669c45292949bfaaa0ff3

supreethmr commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/example/InitiallyOpenShowHide.jsx:
```diff
@@ -20,8 +21,11 @@ class InitiallyOpenShowHide extends React.Component {
+      <ShowHide focusRef={this.focusRef} preview={<p>{previewText}</p>} isOpen={this.state.isOpen} onChange={this.toggleShowHide}>
+        <p>
+          { previewText }
+          <span ref={this.focusRef}>{ text }</span>
```
is this intended change to use same ref for two different html elements.

adoroshk replied:
The `<span>` node created on line 27 needs to be accessed in the `<ShowHide>` component in order to receive focus, so `<ShowHide>` is taking a focusRef prop, which will the reference to the `<span>`. 

I understand that it's confusing as I don't use ref forwarding. The reason for that is following:
- The ref variable and `span` node and both created at the same component;
- The prop can have meaningful name, like `focusRef`;
- The prop is not required.
But I am open for suggestions.

sdadn commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -97,3 +188,4 @@ ShowHide.propTypes = propTypes;
 ShowHide.defaultProps = defaultProps;
 
 export default injectIntl(ShowHide);
+export { Paragraph };
```
Why is this being exported?

sdadn commented on packages/terra-show-hide/src/_ShowHideParagraph.jsx:
Is this an internal React component or something available to consumers to use?

adoroshk replied:
As that component is needed only with ShowHide, the intention was for customers to import it from ShowHide. So I made it internal, import to ShowHide, then export from there as Paragraph (will be renamed per comments below, for now the suggestion is ShowHideFocuser) like this:
import ShowHide, { Paragraph } from 'terra-show-hide';
But I am open for better solutions, more elegant solutions or more consistent solutions.

adoroshk replied:
Also mentioned here: https://github.com/cerner/terra-core/pull/3838/#discussion_r1282362828

eawww reviewed (commented):
In addition to my other comments, I'm getting a very strange behavior while testing in JAWS and Edge. Steps to reproduce:
1. Load dev deployment
2. Navigate to ShowHide About page
3. Use H key to navigate down to examples
4. Step through content of first example with arrow keys.
5. Click the Show button on the first example and continue to step through down to the second example.
6. Step through the preview of the second example.
7. Press the Show button on the second example.
8. **JAWS begins reading content from the first example**

Continuing to other examples seem to work fine.

I'm not sure what's causing this but it could be a real issue. @mjpalazzo are you finding similar behavior?

adoroshk commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +45,120 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+    const [isOpen, setIsOpen] = useState(false);
+    const toggleShowHide = () => setIsOpen(!isOpen);
+
+    // Split text in the preview and the rest of the text portions.
```
@eawww I can reproduce that. It only happens in <li> example, the JAWS gets confused with the previous example if it is open, but for list example only. Please test the list example with the previous example close for now, I am working on figuring out what causes this bug and how to avoid it.

adoroshk replied:
Final fix for JAWS and VO confusion between multiple examples: https://github.com/cerner/terra-core/pull/3838/commits/ef2f17da26d15791ecac86c8dde5f0b911c768d4

adoroshk commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +45,120 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+
+<Notice variant="important" ariaLevel="3">
+
+In order for the component to be accessible, it's important to follow the inplimentation guidance. Even though the focusRef prop is not required for backward compatibility, a failure to use it will result in lack of accessibility.
```
Fixed in https://github.com/cerner/terra-core/pull/3838/commits/0156351be0b642aea10b4a80d2c9297b2d074298

adoroshk commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +45,120 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+            <li>{list[0]}</li>
+            <li>{list[1]}</li>
+            <li>
+              <Paragraph ref={focusRef} text={list[2]} />
```
No, list[3] won't be visible. The showHide component was initially designed in such a way, that only what passed as a preview prop will be visible. In this case the preview portion is header, list[1] and list[2]:
https://github.com/cerner/terra-core/blob/74096b11bbe9ed5372617a339128bc531b7bac2b/packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx#L106-L112

adoroshk commented on packages/terra-show-hide/src/_ShowHideParagraph.jsx:
```diff
@@ -0,0 +1,59 @@
+  return [first, second];
+};
+
+const Paragraph = forwardRef((props, ref) => {
```
I'll take care of that at the very end, once the exact functionality of the component is known, as we still might change something addressing the existing issue.

adoroshk commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +45,120 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+#### Multiple children
+  - At the top level of your component declare a `ref`, pass it as `focusRef` props to `ShowHide` component. Pass a preview node as a prop to `ShowHide` component.
+  - A `Paragraph` component allows `ShowHide` component to set focus on the beginning of the hidden portion of the text upon opening. 
+  - Dublicate the preview nodes as `ShowHide` children.
```
Fixed in https://github.com/cerner/terra-core/commit/0156351be0b642aea10b4a80d2c9297b2d074298

mjpalazzo commented:
@eawww - I also observed some odd behavior. For me, JAWS navigated and announced the default example correctly. I navigated to the second example. JAWS announced list of of 2 items and then annouced the 2 displayed bullet text correctly. I navigated to the Show 2 more sentences example and pressed Enter. JAWS announced "list with 4 items" [which is incorrect] and then the next sentence. I pressed down arrow so JAWS would read the next sentence. I pressed down arrow and navigated to the button. I navigated to the third example. JAWS announced the text of this example. I pressed the button to disclose the hidden text. I pressed down arrow and JAWS announced the disclosed text. I pressed down arrow to navigate to the Hide button. I pressed Enter to activate the button. Here is where it got strange. JAWS announced the 4 bullet of the next example, "bullet Promotes cardiac disease management through health maintenance protocols." It did this before I had entered the example.
I navigated to the next example

adoroshk commented:
@mjpalazzo I assume the problem that you described shows up when the focus needs to be placed within a list item. I updated the examples to avoid that and tested as much as I understand the testing process. Would appreciate if you test as well. If the problem is gone, I will update the docs to warn users not to break in the middle of the list.

mjpalazzo commented:
I suggest the following wording changes:
To make the Show/Hide component accessible, you must use the focusRef prop. This prop allows assistive technologies such as screen readers to put focus on the hidden text and announce it when it is disclosed. The focusRef is optional for backward compatibility but is required to comply with accessibility requirements.

adoroshk commented:
@mjpalazzo This is great, text edited here: https://github.com/cerner/terra-core/pull/3838/commits/0db375b7420ebedcf8deeeb26772aa1f5b651694
<img width="1430" alt="Screenshot 2023-07-28 at 11 56 53 AM" src="https://github.com/cerner/terra-core/assets/119358186/00880cc9-dcb0-4377-a354-e94be30f2a86">

mjpalazzo commented:
@adoroshk - The issues with the examples that I reported previously are resolved. Yeah! Unfortunately I found an issue I had not seen before. Here is what I did:
I navigated to the No Preview Show/Hide Example, pressed Down Arrow to get the Show More button, pressed Enter. JAWS announced all these steps as I did them and then announced the disclosed text. I pressed Down Arrow and went to the Hide button. I pressed enter and the disclosed text collapsed and JAWS immediately began reading the text of the next example starting at "Successful value-based care strategies must demonstrate proficiency in optimizing venues of care to improve health outcomes, lower costs and increase consumer access and satisfaction." It skipped right past the example heading and the Note. I suggest removing the note from the example area and adding some text to the header. Here is a suggestion No focusRef ShowHide (see Accessibility note for information about the focusRef prop).

mjpalazzo commented:
I did some testing with Safari/VoiceOver and had some issues:

- After announcing the preview text VoiceOver does not navigate to the Show/Hide button when I press the Down Arrow. In all examples tested VO skipped the Show/Hide control and went to the Code button. 
- In the ShowHide with Link example, VO announces the link text and instructions for interacting with it and then stops and does not announce the remainder of the sentence. It skipped the rest of the sentence and announced the next sentence.
- In the No Preview Show Hide, VO got lost announcing the preview text. I navigated to the Show More button and pressed Entern to disclose the hidden text. VO announced "CareAware is Cerners comprehensive enterprise-wide solution for device connectivity. CareAware® is a suite of solutions that enable any area of a hospital" and announced "you are on a text element." I pressed Down Arrow and VO announced "hospital" again. I pressed Down Arrow again and VO skipped the text, "to create an" and went directly to "environment where all devices are integrated and confocusableTextually aware to ensure the right data is presented in the right format at the right time."

smason0 commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
````diff
@@ -42,9 +46,113 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+import ShowHide, { Focuser } from 'terra-show-hide';
 ```
 
+## Accessibility
````
Wonder if this section would be better placed after the "Examples" section that explains the basic usage.

adoroshk replied:
Thank you, I am also not sure where to place that block. We might want to re-visit that question after the rest of documentation is approved.

smason0 commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +46,113 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+#### Multiple children
+  Even though the main purpose of the `ShowHide` component is to truncate long paragraps of texts, the content of the `ShowHide` does not have to be one paragraph. It can include links, lists, and any king of HTML tags as long as the first line of the hidden text is passed as a string prop to `Focuser` component rather than an HTML element. Please note:
+  - Links can be placed inside the `p` tag that wrapps `Focuser` component both preceding and following it. The prop `prefix` can be omitted if content has been passes as an HTML element preceding `Focuser`.
+  - Everythin that precedes `Focuser` component must be also included into the `ShowHide` preview prop.
```
sp: "Everything"

adoroshk replied:
Good catch, appreciate it. Fixed in https://github.com/cerner/terra-core/pull/3838/commits/09d4e4543413b7aa554d30b921cfe7f1ec1455b4

eawww commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +46,113 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+
+#### Truncate a paragraph of text
+  - At the top level of your component declare a `ref`, pass it as `focusRef` props to `ShowHide` component. 
+  - Pass a preview node with the portion of the text as a prop to `ShowHide` component.
```
```suggestion
  - Pass a node with the preview portion of the text as a prop to `ShowHide` component.
```

eawww commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +46,113 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+#### Truncate a paragraph of text
+  - At the top level of your component declare a `ref`, pass it as `focusRef` props to `ShowHide` component. 
+  - Pass a preview node with the portion of the text as a prop to `ShowHide` component.
+  - Use `Focuser` component inside `<p>` tag. A `Focuser` component allows `ShowHide` component to set focus on the beginning of the focusable text upon opening. 
```
```suggestion
  - Use the `Focuser` component inside `<p>` tag. A `Focuser` component allows `ShowHide` component to set focus on the beginning of the focusable text upon opening. 
```

eawww commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
````diff
@@ -42,9 +46,113 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+  ```
+
+#### Multiple children
+  Even though the main purpose of the `ShowHide` component is to truncate long paragraps of texts, the content of the `ShowHide` does not have to be one paragraph. It can include links, lists, and any king of HTML tags as long as the first line of the hidden text is passed as a string prop to `Focuser` component rather than an HTML element. Please note:
+  - Links can be placed inside the `p` tag that wrapps `Focuser` component both preceding and following it. The prop `prefix` can be omitted if content has been passes as an HTML element preceding `Focuser`.
````
```suggestion
  Even though the main purpose of the `ShowHide` component is to truncate long paragraps of texts, the content of the `ShowHide` does not have to be one paragraph. It can include links, lists, and any kind of HTML tags as long as the first line of the hidden text is passed as a string prop to `Focuser` component rather than an HTML element.
  
Please note:
  - Links can be placed inside the `p` tag that wraps `Focuser` component both preceding and following it. The prop `prefix` can be omitted if content has been passed as an HTML element preceding `Focuser`.
```

eawww commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +46,113 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+  Even though the main purpose of the `ShowHide` component is to truncate long paragraps of texts, the content of the `ShowHide` does not have to be one paragraph. It can include links, lists, and any king of HTML tags as long as the first line of the hidden text is passed as a string prop to `Focuser` component rather than an HTML element. Please note:
+  - Links can be placed inside the `p` tag that wrapps `Focuser` component both preceding and following it. The prop `prefix` can be omitted if content has been passes as an HTML element preceding `Focuser`.
+  - Everythin that precedes `Focuser` component must be also included into the `ShowHide` preview prop.
+  - Do not break lists placing `Focuser` inside `li`, as it might break the assistive technologies reading order.
```
```suggestion
  - Do not break lists placing `Focuser` inside `li`, as it might cause assistive technologies to read the content incorrectly.
```

eawww commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +46,113 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+</Notice>
+
+#### Truncate a paragraph of text
+  - At the top level of your component declare a `ref`, pass it as `focusRef` props to `ShowHide` component. 
```
```suggestion
  - At the top level of your component declare a `ref`, pass it as `focusRef` prop to `ShowHide` component. 
```

eawww commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +46,113 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+
+</Notice>
+
+#### Truncate a paragraph of text
```
Even though the Notice above gives an h3, this feels like these headings should be level 3 as well since the header in the notice should only apply to what's grouped within it.

adoroshk replied:
Thank you! Changed the headings level to be level 3: https://github.com/cerner/terra-core/pull/3838/commits/09d4e4543413b7aa554d30b921cfe7f1ec1455b4

eawww reviewed (commented):
First pass review for wording/typos. I still need to dive a little deeper.

adoroshk commented:
> @adoroshk - The issues with the examples that I reported previously are resolved. Yeah! Unfortunately I found an issue I had not seen before. Here is what I did: I navigated to the No Preview Show/Hide Example, pressed Down Arrow to get the Show More button, pressed Enter. JAWS announced all these steps as I did them and then announced the disclosed text. I pressed Down Arrow and went to the Hide button. I pressed enter and the disclosed text collapsed and JAWS immediately began reading the text of the next example starting at "Successful value-based care strategies must demonstrate proficiency in optimizing venues of care to improve health outcomes, lower costs and increase consumer access and satisfaction." It skipped right past the example heading and the Note. I suggest removing the note from the example area and adding some text to the header. Here is a suggestion No focusRef ShowHide (see Accessibility note for information about the focusRef prop).

I removed the note and changed the header: https://github.com/cerner/terra-core/pull/3838/commits/e8c4f6d7965e99bac3f627a35b290a441f402073

adoroshk commented:
> I did some testing with Safari/VoiceOver and had some issues:
> 
>     * After announcing the preview text VoiceOver does not navigate to the Show/Hide button when I press the Down Arrow. In all examples tested VO skipped the Show/Hide control and went to the Code button.
> 
>     * In the ShowHide with Link example, VO announces the link text and instructions for interacting with it and then stops and does not announce the remainder of the sentence. It skipped the rest of the sentence and announced the next sentence.
> 
>     * In the No Preview Show Hide, VO got lost announcing the preview text. I navigated to the Show More button and pressed Entern to disclose the hidden text. VO announced "CareAware is Cerners comprehensive enterprise-wide solution for device connectivity. CareAware® is a suite of solutions that enable any area of a hospital" and announced "you are on a text element." I pressed Down Arrow and VO announced "hospital" again. I pressed Down Arrow again and VO skipped the text, "to create an" and went directly to "environment where all devices are integrated and confocusableTextually aware to ensure the right data is presented in the right format at the right time."

@mjpalazzo:
1. I thought it's easily fixable, but I was wrong. I was trying to find a good example of a button which would not present following behavior, and can't see what the difference is between this button and a "code" button. I just observed, that when I walk with Down Arrow on Safari, it is not possible to navigate to buttons in Button component examples or Toggle/Toggle button components. So I am a bit confused. Is that not the expected behavior?
2. Unfortunately it's VO behavior in Safari after something (like line in the text) got focused in the middle of the text and then an Arrow down was clicked - it would just go to the next line, skipping the rest of the line. You can observe the same behavior on https://www.w3.org/WAI/ if via tab navigate to the link inside the text paragraph so that it gets focus (similarly to out link getting focus on opening), then click Down Arrow and observe VO skipping the rest of the line and announcing the next line.
3. This problem comes as a result of issue #2 - after something got focus and then Down Arrow was pressed - it skips the rest of the line, but this time it repeats the last word. It is something that I can't fix for VO as it comes from it's normal behavior. It works as expected with VO + Arrow Rigth though.

adoroshk commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +46,113 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+  - At the top level of your component declare a `ref`, pass it as `focusRef` props to `ShowHide` component. 
+  - Pass a preview node with the portion of the text as a prop to `ShowHide` component.
+  - Use `Focuser` component inside `<p>` tag. A `Focuser` component allows `ShowHide` component to set focus on the beginning of the focusable text upon opening. 
+  - The text splitted in the prefix and focusable text portions has to be passed as string props. The prefix prop should contain the portion of the text that was a part of the preview prop, passed to `ShowHide`.
```
Good catch, appreciate it. Fixed in https://github.com/cerner/terra-core/pull/3838/commits/09d4e4543413b7aa554d30b921cfe7f1ec1455b4

adoroshk commented:
UPD on VO+Safari 

Adding unique id to the focus-able <span> solved the problem of VO getting confused if multiple examples open at the same time. I hoped that might help the issue of focusing within <li> as well, but without display:block property assigned to <li> the focusing still doesn't work correctly, hence the recommendation for customers not truncate the text in the middle of the list stays.

The component has been tested with VO in Safari for being walk-able with Down Arrow. The behavior of the portion of the text after obtaining the focus is consistent with VO behavior when on focused link. Following has been observed:

1. Once the UP/Down arrow is pressed, if the link or span in focus is long enough, the VO sometimes announces the last (for Down Arrow) or first (for up Arrow) word of the focused link/span text. To reproduce this behavior go to https://www.w3.org/WAI/, tab to a link with longer text placed inside the paragraph and press Arrow Up. A first word of a long link text will be announced. 

2. Once the Down arrow is pressed on the focused link or focus-able span, the VO announces the next line of the text, which (unlike the next line with JAWS) doesn't start right after the link/span text ends, but rather from the where the browser-determined next line starts on the screen, meaning that the portion of the text right after the link/span on focus gets skipped. To reproduce this behavior go to https://www.w3.org/WAI/, tab to a link placed inside the paragraph with at least one line of text after the link in that paragraph, then press Arrow Down while on that link. The next line of the text will be announced from the beginning of the string, skipping the rest of the text on the string containing the link.

As this behavior is coming from VO and can't be fixed on the application end, it has been decided to add a Caution notice in Accessibility Section. The placeholder for the notice has been added, I will update it with the actual text once it's provided.

@eawww @mjpalazzo, at that moment the component has been tested by me in following scenarios:
- Microsoft Edge + JAWS, Safari, Chrome, Firefox + VO: tested for being walk-able with TAB and VO + Right/Left Arrow.
- JAWS + Microsoft Edge: tested the show-hide content to be walk-able with Up/Down Arrow.
- Safari + VO: tested the content to be walk-able via Up/Down Arrow.

eawww commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -1,19 +1,28 @@
-Show Hide Component that will show a preview of content and then expand it with a Show More button.
+Show Hide Component - is a disclosure widget that enables content to be either collapsed (hidden) or expanded (visible). It has two elements: a disclosure [button][1] and a section of content whose visibility is controlled by the button. When the controlled content is hidden, the button is often styled as a typical push button with a down-pointing caret to hint that activating the button will dis ...
+
+Another Terra component, [Toggle Button][2], is available and preferred to Show/Hide because it better adheres to accessibility best practices for disclosure components. It differs from Show Hide by displaying the content below the button when it is activated. 
```
```suggestion
Another Terra component, [Toggle Button][2], is available and preferred to Show/Hide when possible because it better adheres to accessibility best practices for disclosure components. It differs from Show Hide by displaying the content below the button when it is activated. 
```

eawww commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +51,127 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+    const [isOpen, setIsOpen] = useState(false);
+    const toggleShowHide = () => setIsOpen(!isOpen);
+
+    const prefix = 'Patients are requesting greater affordability and efficiency in healthcare. With procedures performed in an ambulatory surgery center costing up to 60% less compared to a hospital outpatient department (1), the demand for these facilities is increasing.';
```
```suggestion
    const prefix = 'Patients are requesting greater affordability and efficiency in healthcare. With procedures performed in an ambulatory surgery center costing up to 60% less compared to a hospital outpatient department (1), the demand for these facilities is increasing...';
```

eawww commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +51,127 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+        <ShowHide focusRef={focusRef} preview={preview} isOpen={isOpen} onChange={toggleShowHide} >
+          <h3>{header}</h3>
+          <p>
+            <ShowHideFocuser ref={this.focusRef} prefix={prefix} focusableText="Several factors help bring clarity to life sciences, researchers and clinicians and " />
```
this could use a comment describing exactly what the resulting behavior of this would be.

adoroshk replied:
Adde in https://github.com/cerner/terra-core/pull/3838/commits/1813b072e8074cbe8d2ec0d737f42a88b9b81774

eawww reviewed (approved):
Thank for all the work you've put into this! Very nice job.
A few minor suggestions but over all it looks good. Works great in JAWS.

mjpalazzo commented:
@adoroshk - the Show/Hide is looking good. There is still one navigation problem with the examples and some minor typos:
Custom Button Text Show Hide example - button label has a misspelling: "Composit" should be "Composite"
No Preview Show Hide example - JAWS works fine when I expand the text. When I click the Hide button, JAWS immediately starts reading the text of the next example. After it reads the preview text, I press Down Arrow and focus goes to the Code button above. I press Down Arrow again and JAWS reads the heading of the No focusRef ShowHide example. There is also a typo in the hidden text: "confocusableTextually" probably should be "contextually

adoroshk commented:
The bug in NO Preview example was observed in JAWS version 2023.2302.15, in more recent versions of JAWS 2023.2306.36 the example works as expected.

cm9361 commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -55,9 +67,37 @@ const ShowHide = (props) => {
+    }
+  }, [isOpen, focusRef, contentRef]);
+
+  useEffect(() => {
+    if (containerIsActive && isOpen && contentRef?.current) {
+      contentRef?.current?.focus();
+    }
+  }, [containerIsActive, isOpen]);
```
Why can't this be combined with the previous useEffect?  Couldn't you just set the focus in the else if block for contentRef?

cm9361 commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -75,14 +115,31 @@ const ShowHide = (props) => {
-        {children}
-      </Toggle>
+      <div
+        id={contentId}
```
You should not need to specify an id attribute.

adoroshk replied:
The id here is needed for aria-controls attribute which is part of UXPLATFORM-9055:

https://github.com/cerner/terra-core/blob/67dcc73138ea81f7db11259eff5e3b952e955937/packages/terra-show-hide/src/ShowHide.jsx#L142

cm9361 commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -55,9 +67,37 @@ const ShowHide = (props) => {
+    }
+  }, [containerIsActive, isOpen]);
+
+  const id = uuidv4();
```
I don't think you need to worry about pulling in this new library to generate a unique id.

adoroshk replied:
I thought that is what we were doing recently for uuid, but feel free to show me the example of how we are doing that lately. @sdadn - your input is also very appreciated as for what approach to choose.

adoroshk replied:
renamed to uniqueId in https://github.com/cerner/terra-core/pull/3838/commits/6ad93a1495c9721c4d1f7f0d046706d56c764413

cm9361 commented on packages/terra-show-hide/src/ShowHide.module.scss:
```diff
@@ -14,4 +22,12 @@
+  .show-hide-content {
+    &:focus,
+    :focus {
+      outline: var(--terra-show-hide-focus-outline, 1px dashed);
```
Isn't this different than our normal border style for focus?

adoroshk replied:
Yes, it was provided by @eawww

cm9361 commented on packages/terra-show-hide/src/ShowHideFocuser.jsx:
```diff
@@ -0,0 +1,89 @@
+
+const propTypes = {
+  /**
+   * Text that precedes the focusable portion.
```
If the prop is focusableText, is this comment correct.  The comment states that it is the text that precedes the focusable portion.

adoroshk replied:
Good catch, it was actually the other way around. Fixed in https://github.com/cerner/terra-core/pull/3838/commits/c8f17c023738231996cc1c6baa69f7c03d4a046b

cm9361 commented on packages/terra-show-hide/src/ShowHideFocuser.jsx:
```diff
@@ -0,0 +1,89 @@
+    return null;
+  }
+  const text = string.trim();
+  const maxLength = 155;
```
Is this value defined somewhere as the maximum for assistive technologies?  I am not seeing this value when doing a Google search.

adoroshk replied:
It is not documented anywhere, I had to experiment with JAWS to determine that number.

cm9361 commented on packages/terra-show-hide/src/ShowHideFocuser.jsx:
```diff
@@ -0,0 +1,89 @@
+ * @param str - string value
+ * @returns a trimmed string with trailing white space added
+ */
+const addTrailingWhiteSpace = (str) => `${str.trim()} `;
```
We should have a better variable name than "str".

adoroshk replied:
Agreed, I'll rename it and update this comment with the commit # once it's done

adoroshk replied:
> We should have a better variable name than "str".

I renamed it to string as it is for basically any string value and I had hard time finding better option, but feel free to suggest a better name: https://github.com/cerner/terra-core/pull/3838/commits/c8f17c023738231996cc1c6baa69f7c03d4a046b, but feel free to suggest a better name.

cm9361 commented on packages/terra-show-hide/src/ShowHideFocuser.jsx:
```diff
@@ -0,0 +1,89 @@
+  if (text?.length < maxLength) {
+    return [text];
+  }
+  const lastFullWordEndPosition = text.slice(0, maxLength).lastIndexOf(' ');
```
What happens if the 156th character is the space?

adoroshk replied:
It is ok if the 156 is the space. The next two lines here would trim the spaces:
https://github.com/cerner/terra-core/blob/67dcc73138ea81f7db11259eff5e3b952e955937/packages/terra-show-hide/src/ShowHideFocuser.jsx#L47-L48
If the second portion is an empty string it will just add an empty sting at the end of the span with text, which is totally fine:
[      {truncatedText[1] ? addHeadingWhiteSpace(truncatedText[1]) : null}](https://github.com/cerner/terra-core/blob/67dcc73138ea81f7db11259eff5e3b952e955937/packages/terra-show-hide/src/ShowHideFocuser.jsx#L83)

cm9361 replied:
Your maxlength is 155.  If the 156th character is a space, it won't find the index of that space right?  Won't it then remove this last word?  I am wondering if this boundary is not handled properly.

cm9361 commented on packages/terra-show-hide/src/ShowHideFocuser.jsx:
```diff
@@ -0,0 +1,89 @@
+  const truncatedText = truncate(focusableText);
+  const [focusable, setFocusable] = useState(true);
+
+  const id = uuidv4();
+  const refId = `focus-ref-${id}`;
```
I don't think you will need an id attribute.

cm9361 commented on packages/terra-show-hide/src/ShowHideFocuser.jsx:
```diff
@@ -0,0 +1,89 @@
+        ref={ref}
+        tabIndex={focusable ? '-1' : null}
+        role={focusable ? 'group' : null}
+        onBlur={() => setFocusable(false)}
```
This will create a new method on every rerender of this component.  You should create a named function to handle this instead of using an anonymous function.

cm9361 replied:
How does the component ever get back to a focusable state?  Should this not be allowed?

cm9361 commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -75,14 +115,31 @@ const ShowHide = (props) => {
     }
   }
 
+  const onBlur = () => {
```
What happens if there is a blur event and then the control regains focus?  Does that need to be handled?

adoroshk replied:
The specifics of this component is that the only purpose of the focus is to make the assistive technologies announce the text upon opening. For that we need the component to be only focus-able once, and when it looses focus - lose it forever. The only option for the component to become focus-able able again is if user closes and open it back.

cm9361 commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -97,3 +154,4 @@ ShowHide.propTypes = propTypes;
 ShowHide.defaultProps = defaultProps;
 
 export default injectIntl(ShowHide);
+export { ShowHideFocuser };
```
Why is the ShowHide component exporting the ShowHideFocuser component as well?

adoroshk replied:
The ShowHideFocuser provides additional HTML structure around text to allow JAWS walk throw the text with Down Arrow. The user needs to import it along with ShowHide component from the show-hide package. Exporting ShowHideFocuser from the ShowHide.jsx allows to import it as one line with ShowHide like this:

`import ShowHide, {ShowHideFocuser} from 'terra-show-hide';`
similarly to how we import this:
`import React, {useState} from 'react'; `

adoroshk replied:
@sdadn index file added: https://github.com/cerner/terra-core/pull/3838/commits/822fe0339ce00a4d80c2b649f25b3dc2def71971

mjpalazzo commented:
@adoroshk - Upgrading JAWS to the latest version fixed the problem I observed. :)

sdadn commented on packages/terra-show-hide/tests/jest/ShowHideFocuser.test.jsx:
```diff
@@ -0,0 +1,42 @@
+    return <p><ShowHideFocuser ref={ref} prefix={prefix} focusableText={focusableText} /></p>;
+  };
+
+  afterEach(() => {
+    jest.restoreAllMocks();
+  });
```
`restoreAllMocks` only works with spies.

https://jestjs.io/docs/jest-object#jestrestoreallmocks 

You should use `resetAllMocks` instead. However, that should be done once at the end in the `afterAll` function.

adoroshk replied:
fixed here: https://github.com/cerner/terra-core/pull/3838/commits/aa41c4b79d15cae9e48fece9b1bcbdc462dbe454

sdadn commented on packages/terra-show-hide/tests/jest/ShowHideFocuser.test.jsx:
```diff
@@ -0,0 +1,42 @@
+import React from 'react';
+import ShowHideFocuser from '../../src/ShowHideFocuser';
+
+jest.mock('uuid', () => ({ v4: () => '00000000-0000-0000-0000-000000000000' }));
```
let's move this to a `beforeAll` function so we are setting and reseting them in the same place.

adoroshk replied:
For some reason I can't make it work with beforeAll. I looked for examples and everything I was able to find is in the way it is in this implementation. Can we make this change separately, outside of this PR as it is the way it's done terra-wide?

sdadn replied:
That's fine by me!

sdadn commented on packages/terra-show-hide/tests/jest/ShowHide.test.jsx:
```diff
@@ -5,6 +5,8 @@ import ThemeContextProvider from 'terra-theme-context/lib/ThemeContextProvider';
 import { mountWithIntl } from 'terra-enzyme-intl';
 import ShowHide from '../../src/ShowHide';
 
+jest.mock('uuid', () => ({ v4: () => '00000000-0000-0000-0000-000000000000' }));
```
This mock needs to be reset.

adoroshk replied:
Can you get an example if it's not what's done here: https://github.com/cerner/terra-core/pull/3838/commits/aa41c4b79d15cae9e48fece9b1bcbdc462dbe454

sdadn replied:
I'm okay with doing this in a separate PR as suggested here:
https://github.com/cerner/terra-core/pull/3838/files#r1291684878

sdadn commented on packages/terra-show-hide/package.json:
```diff
@@ -35,7 +35,8 @@
     "terra-theme-context": "^1.0.0",
-    "terra-toggle": "^3.56.0"
+    "terra-toggle": "^3.56.0",
+    "uuid": "7.0.3"
```
We are using uuid `3.4.0` now

adoroshk replied:
Good catch! Fixed in https://github.com/cerner/terra-core/pull/3838/commits/aa41c4b79d15cae9e48fece9b1bcbdc462dbe454

adoroshk commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +51,128 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+
+###  Keyboard Interactions 
+  When the disclosure control has focus:
+  - Enter key activates the disclosure control and toggles the visibility of the disclosure content.
+  - Space key activates the disclosure control and toggles the visibility of the disclosure content. 
```
I guess @mjpalazzo put it that way in purpose. He can answer that better than me.

adoroshk commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
````diff
@@ -42,9 +51,128 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+  ```
+
+### First hidden child is an interactive element (link, button, etc)
+  - At the top level of your component declare a `ref`.
````
fixed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

adoroshk commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +51,128 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+Please note:
+  - Links can be placed inside the `p` tag that wraps `ShowHideFocuser` component both preceding and following it. The prop `prefix` can be omitted if content has been passed as an HTML element preceding `ShowHideFocuser`.
+  - Everything that precedes `ShowHideFocuser` component must be also included into the `ShowHide` preview prop.
+  - Do not break lists placing `ShowHideFocuser` inside `li`, as it might cause assistive technologies to read the content incorrectly.
```
fixed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

adoroshk commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +51,128 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+  Even though the main purpose of the `ShowHide` component is to truncate long paragraps of texts, the content of the `ShowHide` does not have to be one paragraph. It can include links, lists, and any kind of HTML tags as long as the first line of the hidden text is passed as a string prop to `ShowHideFocuser` component rather than an HTML element.
+  
+Please note:
+  - Links can be placed inside the `p` tag that wraps `ShowHideFocuser` component both preceding and following it. The prop `prefix` can be omitted if content has been passed as an HTML element preceding `ShowHideFocuser`.
```
fixed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

adoroshk commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +51,128 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+  
+Please note:
+  - Links can be placed inside the `p` tag that wraps `ShowHideFocuser` component both preceding and following it. The prop `prefix` can be omitted if content has been passed as an HTML element preceding `ShowHideFocuser`.
+  - Everything that precedes `ShowHideFocuser` component must be also included into the `ShowHide` preview prop.
```
fixed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

adoroshk commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -1,19 +1,28 @@
 # Terra Show Hide
 
-Show Hide Component that will show a preview of content and then expand it with a Show More button.
+Show Hide Component - is a disclosure widget that enables content to be either collapsed (hidden) or expanded (visible). It has two elements: a disclosure [button][1] and a section of content whose visibility is controlled by the button. When the controlled content is hidden, the button is often styled as a typical push button with a down-pointing caret to hint that activating the button will dis ...
```
fixed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

adoroshk commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +51,128 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+
+<Notice variant="important" ariaLevel="3">
+
+  To make the Show/Hide component accessible, you must use the `focusRef` prop. This prop allows assistive technologies such as screen readers to move user focus backward to hidden text and announce it when it is disclosed. For backward compatibility, the `focusRef` prop is optional, but is required to meet accessibility requirements.
```
fixed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

adoroshk commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +51,128 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+
+<Notice variant="caution" ariaLevel="3">
+
+  This component is not recommended for use in Safari due to VoiceOver focus limitations with Show Hide. Consider using Terra Toggle Button instead.
```
fixed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

adoroshk commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +51,128 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+  - Enter key activates the disclosure control and toggles the visibility of the disclosure content.
+  - Space key activates the disclosure control and toggles the visibility of the disclosure content. 
+
+### Truncate a paragraph of text
```
Fixed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

adoroshk commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +51,128 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+
+### Truncate a paragraph of text
+
+  - At the top level of your component declare a `ref`, pass it as `focusRef` prop to `ShowHide` component.
```
Fixed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

adoroshk commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
```diff
@@ -42,9 +51,128 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+    const toggleShowHide = () => setIsOpen(!isOpen);
+
+    const prefix = 'Patients are requesting greater affordability and efficiency in healthcare. With procedures performed in an ambulatory surgery center costing up to 60% less compared to a hospital outpatient department (1), the demand for these facilities is increasing.';
+    const focusableText = 'In fact, the U.S. ambulatory surgery center market is expected to see a 6.9 % compound annual growth rate, reaching $33 billion by 2028. (2) Cerner understands the urgency to grow in the ambulatory surgery center market while continuing to deliver excellent care. Healthcare IT products can help improve clinician efficiency and patient outcomes, as well as enhance commun ...
```
removed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

adoroshk commented on packages/terra-core-docs/src/terra-dev-site/doc/show-hide/About.1.doc.mdx:
````diff
@@ -42,9 +51,128 @@ The Show-Hide component must be composed inside the [Base][1] component with a l
+  ```
+
+### Multiple children
+  Even though the main purpose of the `ShowHide` component is to truncate long paragraps of texts, the content of the `ShowHide` does not have to be one paragraph. It can include links, lists, and any kind of HTML tags as long as the first line of the hidden text is passed as a string prop to `ShowHideFocuser` component rather than an HTML element.
````
fixed in https://github.com/cerner/terra-core/pull/3838/commits/a47b7a606f2652ca637d33193c61c798e33195c1

adoroshk commented on packages/terra-show-hide/src/ShowHide.jsx:
```diff
@@ -44,6 +55,7 @@ const defaultProps = {
   buttonAlign: 'start',
   isOpen: false,
   preview: undefined,
```
Updated: https://github.com/cerner/terra-core/blob/2e998ebb9dc3d7bba23f1a29586f8388f3756d52/packages/terra-show-hide/src/ShowHide.jsx#L57

adoroshk commented on packages/terra-show-hide/src/ShowHideFocuser.jsx:
```diff
@@ -0,0 +1,89 @@
+  }
+  const text = string.trim();
+  const maxLength = 155;
+  if (text?.length < maxLength) {
```
removed in https://github.com/cerner/terra-core/pull/3838/commits/822fe0339ce00a4d80c2b649f25b3dc2def71971

adoroshk commented:
@sdadn for some reason I can't answer to the origional comment, so adding it here. Index file added for exports: https://github.com/cerner/terra-core/pull/3838/commits/822fe0339ce00a4d80c2b649f25b3dc2def71971

sdadn commented on packages/terra-show-hide/src/index.jsx:
I think this can be just a `.js` file as we're not using any `.jsx` features.

adoroshk replied:
Oh, that's right, I thought I created a `.js` file. Fixing it, will push a commit in a minute.

adoroshk replied:
Here it is: https://github.com/cerner/terra-core/pull/3838/commits/41cac25858615efd12516f3af81821519c5d0f9f
